File: minimumIsland.js
const minimumIsland = (grid) => {
  const visited = new Set()
  let min = Infinity
  
  for(let r = 0; r < grid.length; r++){
    for(let c = 0; c < grid[0].length; c++){
      const count = explore(grid,r,c,visited)
      if(count) min = Math.min(min, count)
    }
  }
  
  return min
};

const explore = (grid, r, c, visited) => {
  const inRow = 0 <=r && r < grid.length
  const inCol = 0 <= c && c < grid[0].length
  
  if(!inRow || !inCol) return 0
  if(grid[r][c] === 'W') return 0
  
  const pos = `${r},${c}`
  
  if(visited.has(pos)) return 0
  visited.add(pos)
  
  const down = [grid, r + 1, c, visited]
  const up = [grid, r - 1, c, visited]
  const right = [grid, r, c + 1, visited]
  const left = [grid, r, c - 1, visited]
  
  return 1 + explore(...down) + explore(...up) + explore(...left) + explore(...right)
}

File: minimum_island.py
def minimum_island(grid):
  visited = set()
  currMin = float("inf")
  
  for r in range(len(grid)):
    for c in range(len(grid[0])):
      if(pos(r,c) not in visited and grid[r][c] == 'L'):
        currMin = min(currMin, islandSize(grid, r, c, visited))
        
  return currMin

  
  
def pos(r,c):
  return f"{r},{c}"

def inRange(g,r,c):
  return r >= 0 and r < len(g) and c >=0 and c < len(g[0])

def islandSize(grid, r, c, visited):
  if not inRange(grid,r,c) or grid[r][c] == 'W':
    return 0
  
  if pos(r,c) in visited:
    return 0
  
  visited.add(pos(r,c))
  
  size = 1
  
  positions = [
    (r+1,c),
    (r-1,c),
    (r,c-1),
    (r,c+1)
  ]
  
  for row,col in positions:
    size += islandSize(grid, row, col, visited)
  
  return size
      
      